gax/src/main/java/com/google/api/gax/tracing/TracedBidiCallable.java

```diff
@@ -0,0 +1,220 @@
+import com.google.common.base.Preconditions;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nonnull;
+
+/**
+ * A wrapper callable that will wrap a callable chain in a trace.
+ *
+ * <p>This class is meant to be an internal implementation google-cloud-java clients only.
+ */
+@BetaApi("The surface for tracing is not stable and might change in the future")
+@InternalApi("For internal use by google-cloud-java clients only")
+public class TracedBidiCallable<RequestT, ResponseT>
+    extends BidiStreamingCallable<RequestT, ResponseT> {
+
+  @Nonnull private final ApiTracerFactory tracerFactory;
+  @Nonnull private final SpanName spanName;
+  @Nonnull private final BidiStreamingCallable<RequestT, ResponseT> innerCallable;
+
+  public TracedBidiCallable(
+      @Nonnull BidiStreamingCallable<RequestT, ResponseT> innerCallable,
+      @Nonnull ApiTracerFactory tracerFactory,
+      @Nonnull SpanName spanName) {
+    this.tracerFactory = Preconditions.checkNotNull(tracerFactory, "tracerFactory can't be null");
+    this.spanName = Preconditions.checkNotNull(spanName, "spanName can't be null");
+    this.innerCallable = Preconditions.checkNotNull(innerCallable, "innerCallable can't be null");
+  }
+
+  @Override
+  public ClientStream<RequestT> internalCall(
+      ResponseObserver<ResponseT> responseObserver,
+      ClientStreamReadyObserver<RequestT> onReady,
+      ApiCallContext context) {
+
+    ApiTracer tracer = tracerFactory.newTracer(spanName);
+    context = context.withTracer(tracer);
+
+    AtomicReference<Throwable> cancellationCauseHolder = new AtomicReference<>(null);
+
+    ResponseObserver<ResponseT> tracedObserver =
+        new TracedResponseObserver<>(tracer, responseObserver, cancellationCauseHolder);
+    ClientStreamReadyObserver<RequestT> tracedReadyObserver =
+        new TracedClientStreamReadyObserver<>(tracer, onReady, cancellationCauseHolder);
+
+    try {
+      ClientStream<RequestT> clientStream =
+          innerCallable.internalCall(tracedObserver, tracedReadyObserver, context);
+      return new TracingClientStream<>(tracer, clientStream, cancellationCauseHolder);
+    } catch (RuntimeException e) {
+      tracer.operationFailed(e);
+      throw e;
+    }
+  }
+
+  /**
+   * {@link ResponseObserver} wrapper to annotate the current trace with received messages and to
+   * close the current trace upon completion of the RPC.
+   */
+  private static class TracedResponseObserver<ResponseT> implements ResponseObserver<ResponseT> {
+    private final ApiTracer tracer;
+    private final ResponseObserver<ResponseT> innerObserver;
+    private final AtomicReference<Throwable> cancellationCauseHolder;
+
+    private TracedResponseObserver(
+        ApiTracer tracer,
+        ResponseObserver<ResponseT> innerObserver,
+        AtomicReference<Throwable> cancellationCauseHolder) {
+      this.tracer = tracer;
```

Comment:
This seems identical to the Server Streaming PR, but it does not have null pointer check. It is either the null check is redundant there or missing here.

Reply:
I switched to the ResponseObserver from the server streaming PR